lucene/core/src/java/org/apache/lucene/search/Weight.java

```diff
@@ -266,4 +274,45 @@ static void scoreAll(LeafCollector collector, DocIdSetIterator iterator, TwoPhas
     }
   }
 
+  /**
+   * Wraps an internal docIdSetIterator for it to start with docID = -1
+   */
+  protected static class RangeDISIWrapper extends DocIdSetIterator {
+    private final DocIdSetIterator in;
+    private final int min;
+    private final int max;
+    private int docID = -1;
+
+    public RangeDISIWrapper(DocIdSetIterator in, int max) {
+      this.in = in;
+      this.min = in.docID();
+      this.max = max;
+    }
+
+    @Override
+    public int docID() {
+      return docID;
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+      return advance(docID + 1);
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+      target = Math.max(min, target);
+      if (target >= max) {
+        return docID = NO_MORE_DOCS;
+      }
+      return docID = in.advance(target);
```

Comment:
it just occurred to me that this implementation is not correct in the case that the minimum bound of the range of doc IDs to score is less than the current doc ID of the scorer, have you seen any failures with your change? I wonder that we would need to do
```
if (target >= scorer.docID()) { return scorer.docID(); }
```
but we should create a test that fails without this

Reply:
Indeed the [recent failures](https://elasticsearch-ci.elastic.co/job/apache+lucene-solr+master/5689/console) on Lucene are because of this. `ReqExclBulkScorer` can `score` method several times and sometimes with the minimum bound that is <= scorer.docID().



I am thinking the logic is becoming more difficult. I am thinking to go back to the [initial commit](https://github.com/apache/lucene-solr/pull/1952/commits/574245c96db19291138b2eb7253451b567c69d89) and advance collectorIterator to the same doc as `scorerIterator`. At the beginning a  `collectorIterator` matches all docs, so it should precisely advance to the `scorerIterator.docID()).

Reply:
@jpountz  I've created a new PR: https://github.com/apache/lucene-solr/pull/1955.  So sorry for the mess.

Reply:
@jpountz  
> By this `if (target >= scorer.docID()) { return scorer.docID(); }`

Did you mean if (target <= scorer.docID()) { return scorer.docID(); } ?